### DISCUSSION TITLE
Is there an auto way to have sections name assigned to titles?

### DISCUSSION THREAD

**Discussion by aborruso** [quarto-dev/quarto-cli#2825] [Q&A]
Hi,
I have this kind of book chapter

```
# Chapter Name {#sec-chaptername}

## Chapter Section 1

## Chapter Section 2
```

Is there a way to write in another chapter something like `@sec-chaptername#chapter-section-2` to reference a link to the second section of this chapter?

Or the only way is to add a "standard" link as `[my label](my-chapter.qmd#chapter-section-2)`?

Thank you

**Comment by dragonstyle** [comment]
You can use the same `id` style syntax for sections. For example:

```markdown
# Introduction

This is a book created from markdown and executable code.

## Section 1 {#sec-intro-1}

See @knuth84 for additional discussion of literate programming.


See @sec-intro-1
```

**Reply by aborruso** [reply]
I think I have not explained myself. Excuse me.

If I have a single page that contains 30 H2 title , if for some reason I want to reference to one of this 30, I have to add by hand something like {#sec-intro-xx}.
If for same reason I will need to add in the other pages of my site/book a reference to each of those thirty H2, I'll have to add this by hand 30 times, for each each H2 title of that page.

My question is: is it possibile to add a parameter - something like `auto-section-reference` - that for each title create a mnemonically rememberable section ref.
If my title was "Hello world", to have the opportunity to use `@sec-hello_world` (without writing `{#sec-hello_world}`).  Clearly imagining a way to avoid uniqueness problems.

I hope I have explained

Thank you @dragonstyle

**Reply by mcanouil** [reply]
This could indeed be really handy to have an id created, at rendering, for all sections (_i.e._, `h*` tags for HTML).
The possible tricky part is to eventually handle the section level and to ensure ids are unique.

**Reply by cscheid** [reply]
Try this filter:

```lua
local name_tbl = {}

function Header(header)
  local text = pandoc.utils.stringify(header.content)
  text = text:lower():gsub(" ", "-")
  if header.identifier ~= text then
    return
  end
  text = "sec-" .. text
  if name_tbl[text] ~= nil then
    print("warning: repeated id" .. text .. ". Skipping")
  else
    header.identifier = text
    name_tbl[text] = true
    return header
  end
end
```

Minimal example:

````
---
title: Test
filters: 
 - autosectionref.lua
---

## A section

This is a test. see @sec-a-section.

## Another section

Another test.
````

**Reply by dragonstyle** [reply]
^ This is awesome.

One thing to keep in mind about automatically generated section ids is that as the section text changes, the id will change (and references to the section will break). We should warn in this case, making it not so bad to fix, but just FYI.

**Reply by aborruso** [reply]
I will try tomorrow. Thank you @cscheid 

The quarto dev community is really fantastic ❤️

**Reply by aborruso** [reply]
In my opinion it would be great to have it as feature. It's only a proposal, I'm not able to contribute with coding :(
